internal_ws/ykcompile/src/lib.rs

```diff
@@ -146,3 +161,20 @@ impl CompiledTrace {
         self.mc.ptr(dynasmrt::AssemblyOffset(0))
     }
 }
+
+/// Returns a pointer to the static symbol `sym`, or an error if it cannot be found.
+///
+/// In general, there is no guarantee that a symbol will be found, especially when dealing with
+/// binary executables where there's no public API to speak of. In such cases the Rust compiler is
+/// free to (and often does) throw away symbols. Note however that symbols marked `#[no_mangle] are
+/// always exported.
+pub fn find_symbol(sym: &str) -> Result<*mut c_void, CompileError> {
+    let sym_arg = CString::new(sym).unwrap();
+    let addr = unsafe { dlsym(RTLD_DEFAULT, sym_arg.into_raw()) };
+
+    if addr.is_null() {
```

Comment:
A minor stylistic issue which *might* also be a minor optimisation: I assume that the `Ok` branch is the common case? In which case I think it's generally clearer to put it in the `true` branch (which can also play better with optimisers, branch predictors, depending on the phase of the moon etc etc).

Reply:
LLVM optimizes the following function to the same assembly:

```rust
pub fn foo(a: *const u8) -> Result<*const u8, String> {
    if a.is_null() {
        Err(String::new())
    } else {
        Ok(a)
    }
}

pub fn bar(a: *const u8) -> Result<*const u8, String> {
    if !a.is_null() {
        Ok(a)
    } else {
        Err(String::new())
    }
}
```

In fact it merges the functions into one if they are defined in the same object file.

```asm
playground::bar:
	movq	%rdi, %rax
	testq	%rsi, %rsi
	je	.LBB0_1
	movq	%rsi, 8(%rax)
	xorl	%ecx, %ecx
	movq	%rcx, (%rax)
	retq

.LBB0_1:
	movq	$1, 8(%rax)
	xorps	%xmm0, %xmm0
	movups	%xmm0, 16(%rax)
	movl	$1, %ecx
	movq	%rcx, (%rax)
	retq
.set playground::foo, playground::bar
```

I do agree it is a bit nicer though.

Reply:
At this level of macro-optimisation, what *actually* ends up happening is extremely hard to predict, especially with changing versions of the compiler :) I didn't expect it to merge them into one, however!

Reply:
Rust uses the `unnamed_addr` flag for all functions, which explicitly allows this optimization. Search for "unnamed_addr" in https://llvm.org/docs/LangRef.html for more info.

Reply:
@vext01 You might have missed this comment?

Reply:
I thought you'd come to a conclusion to not adjust this. Shall I change it then?

Reply:
For readability if nothing else, having the common case first seems better?

Reply:
Fixed in 3272405017.